File: ios/FiHavenCore/Sources/FiHavenCore/Models/CategoryIcon.swift
import Foundation

/// Resolved category icon — emoji glyph or a small uploaded image (data URI).
/// Mirrors web `categoryIcons.js` (`parseIconValue` / `categoryIconInfo`).
public enum CategoryIcon: Equatable, Sendable, Hashable {
    case emoji(String)
    case image(dataURI: String)

    /// Soft cap matching web `MAX_ICON_DATA_URI_LEN`.
    public static let maxDataURILength = 12_000

    /// Glyph for text-only contexts; images fall back to `defaultEmoji`.
    public func emoji(default defaultEmoji: String = "📌") -> String {
        switch self {
        case .emoji(let e): return e
        case .image: return defaultEmoji
        }
    }

    /// Parse a stored override value (plain emoji string or `{type,value}` object).
    public static func parse(_ raw: JSONValue?) -> CategoryIcon? {
        guard let raw else { return nil }
        if let s = raw.asString {
            return parseEmoji(s)
        }
        guard let obj = raw.asObject else { return nil }
        let type = obj["type"]?.asString
        let value = obj["value"]?.asString?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !value.isEmpty else { return nil }
        if type == "image" {
            return isSafeDataURI(value) ? .image(dataURI: value) : nil
        }
        if type == nil || type == "emoji" {
            return parseEmoji(value)
        }
        return nil
    }

    /// Parse the full `settings.categoryIcons` map.
    public static func parseMap(_ raw: JSONValue?) -> [String: CategoryIcon] {
        guard let obj = raw?.asObject else { return [:] }
        var out: [String: CategoryIcon] = [:]
        for (k, v) in obj {
            if let icon = parse(v) { out[k] = icon }
        }
        return out
    }

    public static func parseEmoji(_ raw: String) -> CategoryIcon? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, trimmed.count <= 16 else { return nil }
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") || trimmed.hasPrefix("data:") {
            return nil
        }
        // Reject plain ASCII labels that aren't icons.
        if trimmed.unicodeScalars.allSatisfy({
            CharacterSet.alphanumerics.contains($0) || $0 == " " || $0 == "_" || $0 == "." || $0 == "-"
        }) {
            return nil
        }
        return .emoji(trimmed)
    }

    public static func isSafeDataURI(_ value: String) -> Bool {
        guard value.count <= maxDataURILength else { return false }
        let lower = value.lowercased()
        return lower.hasPrefix("data:image/png;base64,")
            || lower.hasPrefix("data:image/jpeg;base64,")
            || lower.hasPrefix("data:image/jpg;base64,")
            || lower.hasPrefix("data:image/webp;base64,")
            || lower.hasPrefix("data:image/gif;base64,")
            || lower.hasPrefix("data:image/svg+xml;base64,")
    }
}

File: android/core/src/main/kotlin/app/fihaven/core/model/CategoryIcon.kt
package app.fihaven.core.model

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

/**
 * Resolved icon — emoji glyph, a small uploaded image (data URI), or a
 * bundled issuer brand mark (see `IssuerLogos`).
 * Mirrors web `categoryIcons.js` (`parseIconValue` / `categoryIconInfo`).
 *
 * [parse] only ever produces [Emoji] / [Image]; [Logo] comes from
 * `IssuerIcons`, so a card row and the upcoming list share one renderer.
 */
sealed class CategoryIcon {
    data class Emoji(val value: String) : CategoryIcon()
    data class Image(val dataUri: String) : CategoryIcon()

    /**
     * [key] indexes `IssuerLogos.all`; [emoji] is the stand-in for text
     * contexts and for renderers that can't draw a vector.
     */
    data class Logo(val key: String, val emoji: String) : CategoryIcon()

    /**
     * Issuer initials on a brand-colored chip ([color] is 0xRRGGBB), for
     * issuers with no bundled logo. [emoji] stays the text-context stand-in.
     */
    data class Monogram(val text: String, val color: Int, val emoji: String) : CategoryIcon()

    /** Glyph for text-only contexts; images fall back to [defaultEmoji]. */
    fun emoji(defaultEmoji: String = "📌"): String = when (this) {
        is Emoji -> value
        is Logo -> emoji
        is Monogram -> emoji
        is Image -> defaultEmoji
    }

    companion object {
        const val MAX_DATA_URI_LENGTH = 12_000

        fun parse(raw: JsonElement?): CategoryIcon? {
            if (raw == null) return null
            when (raw) {
                is JsonPrimitive -> return parseEmoji(raw.contentOrNull)
                is JsonObject -> {
                    val type = (raw["type"] as? JsonPrimitive)?.contentOrNull
                    val value = (raw["value"] as? JsonPrimitive)?.contentOrNull?.trim().orEmpty()
                    if (value.isEmpty()) return null
                    if (type == "image") {
                        return if (isSafeDataUri(value)) Image(value) else null
                    }
                    if (type == null || type == "emoji") return parseEmoji(value)
                    return null
                }
                else -> return null
            }
        }

        fun parseMap(raw: JsonElement?): Map<String, CategoryIcon> {
            val obj = raw as? JsonObject ?: return emptyMap()
            val out = mutableMapOf<String, CategoryIcon>()
            for ((k, v) in obj) {
                parse(v)?.let { out[k] = it }
            }
            return out
        }

        fun parseEmoji(raw: String?): CategoryIcon? {
            val trimmed = raw?.trim().orEmpty()
            if (trimmed.isEmpty() || trimmed.length > 16) return null
            if (trimmed.startsWith("http://") || trimmed.startsWith("https://") || trimmed.startsWith("data:")) {
                return null
            }
            if (trimmed.all { it.isLetterOrDigit() || it == ' ' || it == '_' || it == '.' || it == '-' }) {
                return null
            }
            return Emoji(trimmed)
        }

        fun isSafeDataUri(value: String): Boolean {
            if (value.length > MAX_DATA_URI_LENGTH) return false
            val lower = value.lowercase()
            return lower.startsWith("data:image/png;base64,")
                || lower.startsWith("data:image/jpeg;base64,")
                || lower.startsWith("data:image/jpg;base64,")
                || lower.startsWith("data:image/webp;base64,")
                || lower.startsWith("data:image/gif;base64,")
                || lower.startsWith("data:image/svg+xml;base64,")
        }
    }
}
